File: Workshops/Algorithms-BigO-Workshop/exercises/optimize_functions_code.R
# This file contains just the functions from the classify functions exercise. 
# Primarily, this is to make the code easier to run, test, and debug.

# You may wish to modify this code to complete the exercise, or you may wish to 
# create a new file and make your modifications there.


# This function takes in a string representing the genome of an
# organism (or section of a genome) and an integer value k. It 
# returns a list of all the unique substrings of length k
# (often called k-mers) contained in that genome.
unique_kmers <- function(genome, k) {
    kmers <- list()
    final_kmer_position <- nchar(genome) - k + 1
    
    for(start_position in 1:final_kmer_position) {
        # Hint: Consider this line carefully... 
        kmer <- substring(genome, start_position, start_position + k - 1)

        # Looping to check if this kmer is already in the list
        new_kmer <- TRUE
        for(i in seq_along(kmers)) {
            known_kmer <- kmers[i]
            if(known_kmer == kmer) {
                new_kmer <- FALSE
                break
            }
        }

        if(new_kmer) {
            # Note: like above, this is O(n) where n is the length of kmers
            # which may impact your Big O analisys.
            kmers <- append(kmers, kmer)
        }
    }
    
    return(kmers)
}

# This function accepts a string representing a genome
# and an integer k. It returns a dictionary which maps
# each occurring k-mer to the number of times that k-mer
# occurs within the genome.

# Note that it uses the above function, and so you 
# must consider that functions complexity in determining
# this functions complexity.
count_kmers <- function(genome, k) {
    all_kmers <- unique_kmers(genome, k)
    kmer_counts <- list()
    
    for(i in seq_along(all_kmers)) {
        # Start by assuming we don't encounter this kmer
        # (we know we will, but this makes the code cleaner)
        kmer = all_kmers[[i]]
        kmer_counts[[kmer]] <- 0
        
        # Loop over the full genome searching for this kmer
        stop_position <- nchar(genome) - k + 1
        for(start_position in 1:stop_position) {
            inner_kmer <- substring(genome, start_position, start_position + k - 1)

            if(inner_kmer == kmer) {
                kmer_counts[[kmer]] <- kmer_counts[[kmer]] + 1
            }
        }
    }
    
    return(kmer_counts)
}

File: Coding-WorkshopsEtc/b_Sep2020_Algorithms-BigO-Workshop/solutions/optimize_functions_code.py
def unique_kmers(genome, k):
    '''
    This function takes in a string representing the genome of an
    organism (or section of a genome) and an integer value k. It 
    returns a list of all the unique substrings of length k
    (often called k-mers) contained in that genome.

    But the runtime is simply O(GK), and is completely independent of 
    the value U (the number of unique k-mers)
    '''
    kmers = {} # NOTE: a dictionary instead of a list.
    final_kmer_position = len(genome) - k + 1
    
    for start_position in range(final_kmer_position):
        kmer = genome[start_position:start_position + k]

        # We don't even need to check if it's in there or not.
        # If it is, it'll stay true, of it's not it will become true.
        kmers[kmer] = True
    
    return list(kmers.keys())


def count_kmers(genome, k):
    '''
    This function accepts a string representing a genome
    and an integer k. It returns a dictionary which maps
    each occurring k-mer to the number of times that k-mer
    occurs within the genome.

    Instead of finding the unique k-mers and then counting them
    one by one, we just look at the genome once and keep a running
    count of ALL the k-mers at the same time!

    Note that it is almost identical to the above function!
    '''
    kmers = {}
    final_kmer_position = len(genome) - k + 1
    
    for start_position in range(final_kmer_position):
        kmer = genome[start_position:start_position + k]

        if kmer not in kmers:
            kmers[kmer] = 0
        
        kmers[kmer] += 1
    
    return kmers